### ISSUE TITLE
Drawer lag

### ISSUE THREAD

**Issue by kidnapkin** [react-community/react-navigation#1768]
### Current Behavior

On some screens drawer opens extremely slow.
I've recordered current [behavior.](https://youtu.be/7URkkQGXV2s?t=52s)
It's so strange and unexpected for me.

### Expected Behavior

Render drawer correctly everywhere.

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation |  1.0.0-beta.11
| react-native     |  0.44.2
| node             | 7.10
| npm or yarn      | 0.24.5

### Navigator code
```Javascript
import React from 'react'
import {StackNavigator, TabNavigator, DrawerNavigator} from 'react-navigation'
import ChatScreen from '../Containers/ChatScreen'
import styles from './Styles/NavigationStyles'
import {MenuButton} from './MenuButton'
import DrawerContent from '../Containers/DrawerContent'

import SignInScreen from '../Containers/SignInScreen'
import SignUpScreen from '../Containers/SignUpScreen'
import PasswordResetScreen from '../Containers/PasswordResetScreen'
import CurrentUserInfo from '../Containers/CurrentUserInfo'
import CurrentUserServices from '../Containers/CurrentUserServices'
import CurrentUserTravels from '../Containers/CurrentUserTravels'
import IncomingBookingsScreen from '../Containers/IncomingBookingsScreen'
import OutgoingBookingsScreen from '../Containers/OutgoingBookingsScreen'
import ProfileEditScreen from '../Containers/ProfileEditScreen'
import SearchScreen from '../Containers/SearchScreen'
import SearchResultScreen from '../Containers/SearchResultScreen'
import UserInfoScreen from '../Containers/UserProfileInfoScreen'
import UserServicesScreen from '../Containers/UserProfileServicesScreen'
import UserTravelsScreen from '../Containers/UserProfileTravelsScreen'

// Manifest of possible screens
const CurrentUserTabs = TabNavigator({
  Info: {
    screen: CurrentUserInfo
  },
  Services: {
    screen: CurrentUserServices
  },
  Travels: {
    screen: CurrentUserTravels
  }
}, {
  tabBarOptions: {
    scrollEnabled: true,
    animationEnabled: true,
    labelStyle: {
      fontSize: 12
    },
    indicatorStyle: {
      backgroundColor: '#fff'
    },
    style: {
      backgroundColor: '#9b59b6'
    }
  }
})

const UserProfileTabs = TabNavigator({
  UserInfo: {
    screen: UserInfoScreen,
    navigationOptions: {
      title: `Info`
    }
  },
  UserServices: {
    screen: UserServicesScreen,
    navigationOptions: {
      title: `Services`
    }
  },
  UserTravels: {
    screen: UserTravelsScreen,
    navigationOptions: {
      title: `Travels`
    }
  }
}, {
  navigationOptions: {
    headerTitle: 'User profile'
  },
  tabBarOptions: {
    scrollEnabled: true,
    animationEnabled: true,
    labelStyle: {
      fontSize: 12
    },
    indicatorStyle: {
      backgroundColor: '#fff'
    },
    style: {
      backgroundColor: '#9b59b6'
    }
  }
})

const BookingTabs = TabNavigator({
  IncomingBookings: {
    screen: IncomingBookingsScreen,
    navigationOptions: {
      title: 'Incoming Bookings'
    }
  },
  OutgoingBookings: {
    screen: OutgoingBookingsScreen,
    navigationOptions: {
      title: 'Outgoing Bookings'
    }
  }
}, {
  tabBarOptions: {
    scrollEnabled: true,
    animationEnabled: true,
    labelStyle: {
      fontSize: 12
    },
    indicatorStyle: {
      backgroundColor: '#fff'
    },
    tabStyle: {
      width: 200
    },
    style: {
      backgroundColor: '#9b59b6'
    }
  }
})

const Profile = StackNavigator({
  ProfileScreen: {
    screen: CurrentUserTabs,
    navigationOptions: ({ navigation }) => ({
      title: 'Your Profile',
      headerTintColor: '#fff',
      headerLeft: <MenuButton navigation={navigation} />,
      headerStyle: styles.headerTabs
    })
  },
  EditProfile: {
    screen: ProfileEditScreen,
    navigationOptions: {
      title: 'Edit my info',
      headerStyle: styles.header,
      gesturesEnabled: false
    }
  }
}, {
  navigationOptions: {
    headerTintColor: '#fff',
    headerStyle: styles.header
  }
})

const SearchStack = StackNavigator({
  Search: {
    screen: SearchScreen,
    navigationOptions: ({ navigation }) => ({
      title: 'Find native speaker',
      headerStyle: styles.header,
      headerLeft: <MenuButton navigation={navigation} />
    })
  },
  SearchResults: {
    screen: SearchResultScreen,
    navigationOptions: {
      title: 'Search results',
      headerStyle: styles.header
    }
  },
  SearchSingle: {
    screen: UserProfileTabs,
    navigationOptions: {
      headerStyle: styles.headerTabs
    }
  }
}, {
  navigationOptions: {
    headerTintColor: '#fff',
    headerStyle: styles.header
  }
})

const BookingsStack = StackNavigator({
  ProfileBookings: {
    screen: BookingTabs,
    navigationOptions: ({ navigation }) => ({
      headerTitle: 'Your bookings',
      headerLeft: <MenuButton navigation={navigation} />
    })
  },
  UserScreen: {
    screen: UserProfileTabs,
    navigationOptions: {
      headerTitle: 'User Profile'
    }
  }
}, {
  navigationOptions: {
    headerTitle: 'Your bookings',
    headerTintColor: '#fff',
    headerStyle: styles.headerTabs
  }
})

const Main = DrawerNavigator({
  ProfileScreen: {
    screen: Profile
  },
  Search: {
    screen: SearchStack
  },
  Bookings: {
    screen: BookingsStack
  },
  ChatScreen: {
    screen: ChatScreen
  }
}, {
  contentComponent: props => <DrawerContent {...props} />
})

const Auth = StackNavigator({
  SignInScreen: {
    screen: SignInScreen
  },
  SignUpScreen: {
    screen: SignUpScreen
  },
  PasswordResetScreen: {
    screen: PasswordResetScreen
  }
}, {
  navigationOptions: {
    header: null
  }
})

const PrimaryNav = StackNavigator({
  Authentication: {
    screen: Auth
  },
  MainScreen: {
    screen: Main
  }
}, {
  navigationOptions: {
    header: null
  }
})

const initialState = PrimaryNav.router.getStateForAction(Auth.router.getActionForPathAndParams('SignInScreen'))

export const navReducer = (state = initialState, action) => {
  const nextState = PrimaryNav.router.getStateForAction(action, state)
  // Simply return the original `state` if `nextState` is null or undefined.
  return nextState || state
}

export default PrimaryNav
```

**Comment by matthamil** [comment]
The Drawer is very slow for me while I am debugging remotely. Are you debugging remotely while you are experiencing this? Try turning it off and see if it is still slow.

**Comment by kidnapkin** [comment]
It's not a dev mode. It's a release build.

**Comment by LAITONEN** [comment]
I also have lags. In my **simple** test App with the following hierarchy:
```
TabNavigator1:
    View1
    DrawerNavigator:
        TabNavigator2:
            View2
            View3
            View4
         View5
```

The lag with the DrawerNavigator appears when I switch from the View5 to TabNavigator2 through the Drawer being opened by swipe-right.

**# The code:**
### **Reducers:**
```
`import { combineReducers } from 'redux';
import { NavigationActions } from 'react-navigation';

import { AppNavigator } from '../navigators/AppNavigator';
import { MainNavigator } from '../navigators/MainNavigator';
import { AdditionalNavigator } from '../navigators/AdditionalNavigator';

const initialNavAppState = AppNavigator.router.getStateForAction('Login');

// APP NAVIGATOR (TabNavigator1)

function navApp(state = initialNavAppState, action) {
  let nextState;
  switch (action.type) {
    case 'Login':
      nextState = AppNavigator.router.getStateForAction(
        NavigationActions.navigate({ routeName: 'Drawer' }),
        state
      );
      break;
    case 'Logout':
      nextState = AppNavigator.router.getStateForAction(
        NavigationActions.navigate({ routeName: 'Login' }),
        state
      );
      break;
    default:
      nextState = AppNavigator.router.getStateForAction(action, state);
      break;
  }

  return nextState || state;
}

// ADDITIONAL NAVIGATOR (DRAWER)
const initialNavAddState = AdditionalNavigator.router.getStateForAction(NavigationActions.init());


function navAdd(state = initialNavAddState, action) {
    const nextState = AdditionalNavigator.router.getStateForAction(action, state);

  return nextState || state;
}

// MAIN NAVIGATOR (TabNavigator2)
const initialNavMainState = MainNavigator.router.getStateForAction(NavigationActions.init());


function navMain(state = initialNavMainState, action) {
    const nextState = MainNavigator.router.getStateForAction(action, state);

  return nextState || state;
}


// LOGIN / LOGOUT

const initialAuthState = { isLoggedIn: false };

function auth(state = initialAuthState, action) {
  switch (action.type) {
    case 'Login':
      return { ...state, isLoggedIn: true };
    case 'Logout':
      return { ...state, isLoggedIn: false };
    default:
      return state;
  }
}

const AppReducer = combineReducers({
  navApp,
  navAdd,
  navMain,
  auth,
});

export default AppReducer;
```
`

### **TabNavigator1 (AppNavigator):**
```
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addNavigationHelpers, TabNavigator } from 'react-navigation';

import LoginScreen from '../components/LoginScreen';
import AdditionalWithNavigation from './AdditionalNavigator';

export const AppNavigator = TabNavigator({
  Login: { screen: LoginScreen },
  Drawer: { screen: AdditionalWithNavigation }, },
  { navigationOptions: { tabBarVisible: false } }
);

const AppWithNavigationState = ({ dispatch, nav }) => (
  <AppNavigator navigation={addNavigationHelpers({ dispatch, state: nav })} />
);

AppWithNavigationState.propTypes = {
  dispatch: PropTypes.func.isRequired,
  nav: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  nav: state.navApp,
});

export default connect(mapStateToProps)(AppWithNavigationState);

```
### **DrawerNavigator:**
```
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addNavigationHelpers, DrawerNavigator } from 'react-navigation';

import MainWithNavigation from './MainNavigator';
import ProfileScreen from '../components/ProfileScreen';

export const AdditionalNavigator = DrawerNavigator({
  Main: { screen: MainWithNavigation },
  Profile: { screen: ProfileScreen },
});

const AdditionalWithNavigation = ({ dispatch, nav }) => (
  <AdditionalNavigator navigation={addNavigationHelpers({ dispatch, state: nav })} />
);

AdditionalWithNavigation.propTypes = {
  dispatch: PropTypes.func.isRequired,
  nav: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  nav: state.navAdd,
});

export default connect(mapStateToProps)(AdditionalWithNavigation);

```
### **TabNavigator2 (MainNavigator):**
```
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addNavigationHelpers, TabNavigator } from 'react-navigation';

import ScreenOne from '../components/ScreenOne.js';
import ScreenTwo from '../components/ScreenTwo';
import ScreenThree from '../components/ScreenThree';

export const MainNavigator = TabNavigator({
  one: { screen: ScreenOne },
  two: { screen: ScreenTwo },
  three: { screen: ScreenThree },
});

const MainWithNavigation = ({ dispatch, nav }) => (
  <MainNavigator navigation={addNavigationHelpers({ dispatch, state: nav })} />
);

MainWithNavigation.propTypes = {
  dispatch: PropTypes.func.isRequired,
  nav: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  nav: state.navMain,
});

export default connect(mapStateToProps)(MainWithNavigation);
```

**Comment by jayshah123** [comment]
What I have observed is: 

Using Drawer:
Switching to a screen calls render() for that  new screen everytime and screens with lot of content to render take time to be switched to. Ideally render should happen only once when first visiting the screen in drawer.

Using Tabs:
Although some issues have noted that TabNavigator and DrawerNavigator are same behind the scenes,
TabNavigator calls render only once per tab screen, and this should have been the same behaviour for DrawerNavigator as well.

**Comment by jayshah123** [comment]
As a result of putting logs,
Not only render, but screen component is constructed from scratch when screen is switched to, in Drawer.

**Comment by Akshg** [comment]
My drawer was lagging. So I wrote the following code in onItemPress:

```
<DrawerItems 
          {...props}
          onItemPress={({ route, focused }) => {
            if(!focused) {
              setTimeout(() => {              
                props.navigation.navigate(route.routeName)
              }, 0)
            }
            props.navigation.navigate('DrawerClose');
          }}
/>
```
It is navigating, with the help of setTimeout asynchronously.

**Comment by vonovak** [comment]
I'm guessing this is a duplicate of https://github.com/react-community/react-navigation/issues/1395 which was fixed. Feel free to reopen if not.
